### PR TITLE
(graphcache) - expose a way to interact with the client outside of operations

### PR DESCRIPTION
**Pull request by JoviDeCroock**
## Summary

This is very much a draft and experiment whether this is a good idea.

We've repeatedly seen the need for people to update data coming from external sources (socket.io,...) outside of GraphQL so with this PR we want to enable this once and for all. Our stance towards this has been hesitant because of the unintended side-effects that might be created by doing this so this should be considered an escape hatch.

An example would be:

```js
import { useClient, gql } from 'urql';

const TodoFragment = gql`
    fragment todoFields on Todo {
      id
      text
      complete
      __typename
    }
`;

export const Component = () => {
  const client = useClient();
  useEffect(() => {
    myTodoSocket.subscribe((updatedTodo) => {
      cache.updateFragment(TodoFragment, { ...updatedTodo, __typename: 'Todo' }, { id: updatedTodo.id })
    })
  }, [])
}
```

TODO:

- [ ] test whether operations are correctly invoked after updating a query
- [ ] find a way to correctly include `core.d.ts` so client is extended with the cache helper

## Set of changes

- expose `cache` on the client interface
- add helpers so the cache can track deps and data state while updating data

### CODE REVIEW

File: exchanges/graphcache/src/store/data.ts [hunk 1]
```diff
@@ -192,17 +192,7 @@ export const noopDataState = (
...
-    currentOperation !== null,
-    'Invalid Cache call: The cache may only be accessed or mutated during' +
-      'operations like write or query, or as part of its resolvers, updaters, ' +
-      'or optimistic configs.',
-    2
-  );
-
-  return currentOperation;
-};
+export const getCurrentOperation = (): OperationType | null => currentOperation;
```
Comment (kitten) [inline exchanges/graphcache/src/store/data.ts:195]: Nice catch!

File: exchanges/graphcache/src/store/data.ts [hunk 2]
```diff
@@ -414,8 +404,8 @@ export const gc = () => {
 };
 
 const updateDependencies = (entityKey: string, fieldKey?: string) => {
-  if (fieldKey !== '__typename') {
-    if (entityKey !== currentData!.queryRootKey) {
+  if (fieldKey !== '__typename' && currentData) {
```
Comment (kitten) [inline exchanges/graphcache/src/store/data.ts:407]: I think there may be some cases in which this could happen, right? I'm thinking this may be so that `cacheExchange().store.resolve` still works, right? 🤔 Is there any way in which we could ensure that an operation is still available for these cases? (just very loose thoughts)

File: exchanges/graphcache/src/store/store.ts [hunk 3]
```diff
@@ -125,13 +148,63 @@ export class Store<
...
+              this.client!.reexecuteOperation(toRequestPolicy(op, policy));
+            }
+          });
+        }
+        clearDataState();
+      };
+    }
+
+    return null;
+  }
```
Comment (kitten) [inline exchanges/graphcache/src/store/store.ts:191]: This is an interesting one! So, we basically now have a system whereby `store` holds a lot more power over the reexecution behaviour. My gut feeling is that it should be possible for us to get rid of `executePendingOperations` and `collectPendingOperations` entirely somehow, but I think I'll need a closer look at the whole files to see whether that's indeed the case 🤔

**Review by kitten — APPROVED** [review]

File: exchanges/graphcache/src/store/data.ts (hunk 2 above)
Comment (JoviDeCroock) [reply exchanges/graphcache/src/store/data.ts:407]: Hmm, I think this is needed so we don't crash if someone manages to call this without an initialised data state 😅 all though that should not be possible anymore. That would mean we start marking `currentData` as always present which feels a hard one from a typing perspective as it relies on us calling `initDataState`

File: exchanges/graphcache/src/store/store.ts (hunk 3 above)
Comment (JoviDeCroock) [reply exchanges/graphcache/src/store/store.ts:191]: I honestly would prefer that as well, this logic is quite shared currently with the main exchange code.

**Comment by mwangaben** [comment]
This is useful

**Comment by nil1511** [comment]
any updates on this? 🙊

**Comment by paulrostorp** [comment]
@JoviDeCroock could this be merged ?

**Comment by kitten** [comment]
@paulrostorp Last where we left it I had some concerns with how far-reaching this change is. Sorry for not posting it here! Generally, if we merged this as-is, it's a huge departure from imposing one limitation which used to be fundamental: Cache Updates are isolated to the Normalised Cache itself.

This may not seem that big, but prevents a lot of the nuances of _succeeding_ at making normalised caches reliable in GraphQL. We've basically found that if all effects are described by mutations and subscriptions (refetches aside) then it's hard to argue that updates should be possible from the outside.
Generally, what I've seen is that this leads to small exceptions, i.e. small updates being added for UI interactions that don't quite go to a mutation, or aren't quite represented by the GraphQL schema, which can get out of hand rather quickly.

That doesn't mean we won't introduce something like this. We actually had an idea to provide an API whereby "passive" mock subscriptions could be triggered instead. In other words, an API that lets you tell the cache "pretend as if this mutation/subscription came in, and update using it". In all likelihood this is going to fix the original issue here, which was around external event sources, — websockets outside of GraphQL for instance.

That said though, I believe the issue we had with those was that it isn't obvious what the API would look like, and the code to trigger a faux mutation/subscription response _and_ write an updater in Graphcache would be rather excessive.

Another alternative could be for us to somehow encourage having the update logic still be in Graphcache, e.g. by having a "triggers" object in there that can be called from the outside. This was an attractive alternative, since it's like "manual triggered updaters." Yet another alternative was to simply expose less of the `cache` methods to the outside.

However, this still requires us to consider some potential impacts on our "Deterministic Cache Updates" guarantees: https://formidable.com/open-source/urql/docs/graphcache/normalized-caching/#deterministic-cache-updates
Basically, if a synchronous update can be triggered at any time, then it may impact optimistic and non-optimistic cache states equally. In other words, it may not be safe at all times to make a change

So there basically are a couple of hurdles here to making this happen 😅 And they're similar to what we've got for subscriptions in general. However, I think we also got side-tracked by how "unnatural" this feels with our current "dependencies" system, which is the system that determines which operations have updated in the cache.

Once these three questions are solved (1. API design, 2. Deterministic Guarantees, 3. Dependencies System) this is basically a go for us, imho.

**Comment by JoviDeCroock** [comment]
I will close this for the time being 😅

**Comment by paulrostorp** [comment]
@JoviDeCroock Thank you for the detailed answer. I appreciate you taking the responsible approach 🙂 

Unfortunately for me, it means I cannot use this great library 😞

**Comment by JoviDeCroock** [comment]
@paulrostorp mind explaining your use case, it might help us get insight

**Comment by paulrostorp** [comment]
@JoviDeCroock I think has been mentioned before: to update cache from push events outside of subscriptions (in my case WS and https://firebase.google.com/docs/cloud-messaging)
